src/rfid.cpp

```diff
@@ -0,0 +1,460 @@
+#include "rfid.h"
+
+#include "debug.h"
+#include "mqtt.h"
+#include "lcd.h"
+#include "app_config.h"
+#include "input.h"
+#include "openevse.h"
+
+#include <Wire.h>
+
+#ifndef I2C_SDA
+#define I2C_SDA 21
+#endif
+
+#ifndef I2C_SCL
+#define I2C_SCL 22
+#endif
+
+/*
+ * Documentation of the I2C messages:
+ * NXP PN532 manual, https://www.nxp.com/docs/en/user-guide/141520.pdf
+ * (accessed: 2022-01-14)
+ */
+
+#define PN532_I2C_ADDRESS  (0x48 >> 1)
+#define PN532_RDY 0x01
+#define PN532_FRAME_HEADER_LEN 4
+#define PN532_FRAME_IDENTIFIER_LEN 1
+#define PN532_CMD_SAMCONFIGURATION_RESPONSE 0x15
+#define PN532_CMD_INAUTOPOLL_RESPONSE 0x61
+
+#define I2C_READ_BUFFSIZE 35 //might increase later
+
+RfidTask::RfidTask() :
+  MicroTasks::Task()
+{
+}
+
+void RfidTask::begin(EvseManager &evse){
+    _evse = &evse;
+    Wire.begin(I2C_SDA, I2C_SCL);
+    MicroTask.startTask(this);
+}
+
+void RfidTask::setup(){
+    if (config_rfid_enabled()) {
+        pn532_initialize();
+    }
+}
+
+void RfidTask::scanCard(String& uid){
+    if(waitingForTag > 0){
+        waitingForTag = 0;
+        cardFound = true;
+        waitingForTagResult = uid;
+        lcd.display("Tag detected!", 0, 0, 0, LCD_CLEAR_LINE);
+        lcd.display(uid, 0, 1, 3000, LCD_CLEAR_LINE);
+    } else if (onCardScanned && (*onCardScanned) && (*onCardScanned)(uid)){
+        //OCPP will process the card
+    }else{
+        // Check if tag is stored locally
+        char storedTags[rfid_storage.length() + 1];
+        rfid_storage.toCharArray(storedTags, rfid_storage.length()+1);
+        char* storedTag = strtok(storedTags, ",");
+        bool foundCard = false;
+        while(storedTag)
+        {
+            String storedTagStr = storedTag;
+            storedTagStr.replace(" ", "");
+            if(storedTagStr.equals(uid)){
+                foundCard = true;
+                if (!isAuthenticated()){
+                    setAuthentication(uid);
+                    lcd.display("RFID: authenticated", 0, 1, 5 * 1000, LCD_CLEAR_LINE);
+                    DBUGLN(F("[rfid] found card"));
+                } else if (uid == authenticatedTag) {
+                    resetAuthentication();
+                    lcd.display("RFID: finished. See you next time!", 0, 1, 5 * 1000, LCD_CLEAR_LINE);
+                    DBUGLN(F("[rfid] finished by presenting card"));
+                } else {
+                    lcd.display("RFID: card does not match", 0, 1, 5 * 1000, LCD_CLEAR_LINE);
+                    DBUGLN(F("[rfid] card does not match"));
+                }
+                break;
+            }
+            storedTag = strtok(NULL, ",");
+        }
+
+        if (!foundCard) {
+            lcd.display("RFID: did not recognize card", 0, 1, 5 * 1000, LCD_CLEAR_LINE);
+            DBUGLN(F("[rfid] did not recognize card"));
+        }
+
+        // Send to MQTT broker
+        DynamicJsonDocument data{JSON_OBJECT_SIZE(1) + uid.length() + 1};
+        data["rfid"] = uid;
+        mqtt_publish(data);
```

Comment:
Is [this](https://github.com/OpenEVSE/ESP32_WiFi_V4.x/pull/325/files#diff-1518590f5c5028e6a34dd695918cfc679e97e1452dc0dc814718f250d93795deR95) sufficient for that purpose?

Reply:
Not really, that is used for filling in the status when polled or reporting to EmonCMS, `event_send` will send the value over MQTT and the WebSocket so any clients get updated immediately rather than up to 30 seconds later.

Reply:
Not really, that is used for filling in the status when polled or reporting to EmonCMS, `event_send` will send the value over MQTT and the WebSocket so any clients get updated immediately rather than up to 30 seconds later.